runtime/include/wasm_stack.h

```diff
@@ -21,48 +28,93 @@ struct wasm_stack {
  * @returns 0 on success, -1 on error
  */
 static inline int
-wasm_stack_allocate(struct wasm_stack *stack, size_t capacity)
+wasm_stack_init(struct wasm_stack *self, size_t capacity)
 {
-	assert(stack);
+	assert(self);
 
 	int rc = 0;
 
-	stack->buffer = (uint8_t *)mmap(NULL, /* guard page */ PAGE_SIZE + capacity, PROT_NONE,
-	                                MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
-	if (unlikely(stack->buffer == MAP_FAILED)) {
+	self->buffer = (uint8_t *)mmap(NULL, /* guard page */ PAGE_SIZE + capacity, PROT_NONE,
+	                               MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+	if (unlikely(self->buffer == MAP_FAILED)) {
 		perror("sandbox allocate stack");
 		goto err_stack_allocation_failed;
 	}
 
-	stack->low = (uint8_t *)mmap(stack->buffer + /* guard page */ PAGE_SIZE, capacity, PROT_READ | PROT_WRITE,
-	                             MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED, -1, 0);
-	if (unlikely(stack->low == MAP_FAILED)) {
+	self->low = (uint8_t *)mmap(self->buffer + /* guard page */ PAGE_SIZE, capacity, PROT_READ | PROT_WRITE,
+	                            MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED, -1, 0);
+	if (unlikely(self->low == MAP_FAILED)) {
 		perror("sandbox set stack read/write");
 		goto err_stack_prot_failed;
 	}
 
-	stack->capacity = capacity;
-	stack->high     = stack->low + capacity;
+	ps_list_init_d(self);
+	self->capacity = capacity;
+	self->high     = self->low + capacity;
 
 	rc = 0;
 done:
 	return rc;
 err_stack_prot_failed:
-	rc = munmap(stack->buffer, PAGE_SIZE + capacity);
+	rc = munmap(self->buffer, PAGE_SIZE + capacity);
 	if (rc == -1) perror("munmap");
 err_stack_allocation_failed:
-	stack->buffer = NULL;
-	rc            = -1;
+	self->buffer = NULL;
+	rc           = -1;
 	goto done;
 }
 
 static inline void
-wasm_stack_free(struct wasm_stack *stack)
+wasm_stack_free(struct wasm_stack *self)
 {
-	assert(stack != NULL);
-	assert(stack->buffer != NULL);
+	free(self);
+}
+
+
+static struct wasm_stack *
+wasm_stack_new(size_t capacity)
+{
+	struct wasm_stack *self = wasm_stack_allocate();
```

Comment:
I think this demonstrates why the common `_alloc` allocates and calls `_init`, and `_init` initializes the structure is useful. I don't think that your `_allocate` is called in places not followed by `_init`, so it might as well do both.

Reply:
Reading this after writing my other comment above. I think that this change sounds fine and likely addresses your concerns about using a name that is also a C++ keyword.

Reply:
I'll make this change. alloc will wrap init. free will wrap deinit.